File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter17/PrintableTomKitten/PrintableTomKitten/CustomPageRange.cs
using System;
using System.Collections.Generic;

namespace PrintableTomKitten
{
    public class CustomPageRange
    {
        // Structure used internally 
        struct PageRange
        {
            public PageRange(int from, int to) : this()
            {
                this.From = from;
                this.To = to;
            }

            public int From { private set; get; }
            public int To { private set; get; }
        }

        public CustomPageRange(string str, int maxPageNumber)
        {
            List<PageRange> pageRanges = new List<PageRange>();
            string[] strRanges = str.Split(',');

            foreach (string strRange in strRanges)
            {
                int dashIndex = strRange.IndexOf('-');

                // Just one page number
                if (dashIndex == -1)
                {
                    int page;

                    if (Int32.TryParse(strRange.Trim(), out page) &&
                        page > 0 && page <= maxPageNumber)
                    {
                        pageRanges.Add(new PageRange(page, page));
                    }
                    else
                    {
                        return;
                    }
                }
                // Two page numbers separated by a dash
                else
                {
                    string strFrom = strRange.Substring(0, dashIndex);
                    string strTo = strRange.Substring(dashIndex + 1);
                    int from, to;

                    if (Int32.TryParse(strFrom.Trim(), out from) &&
                        Int32.TryParse(strTo.Trim(), out to) &&
                        from > 0 && from <= maxPageNumber && 
                        to > 0 && to <= maxPageNumber &&
                        from <= to)
                    {
                        pageRanges.Add(new PageRange(from, to));
                    }
                    else
                    {
                        return;
                    }
                }
            }

            // If we made it to this, the input string is valid
            this.PageMapping = new List<int>();

            // Define a mapping to page numbers
            foreach (PageRange pageRange in pageRanges)
                for (int page = pageRange.From; page <= pageRange.To; page++)
                    this.PageMapping.Add(page);
        }

        // Zero-based in, one-based out
        public IList<int> PageMapping { private set; get; }

        public bool IsValid 
        {
            get { return this.PageMapping != null; }
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter17/PrintableTomKitten/PrintableTomKitten/CustomPageRange.cpp
#include "pch.h"
#include "CustomPageRange.h"

using namespace PrintableTomKitten;

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;

CustomPageRange::CustomPageRange(String^ str, int maxPageNumber)
{
    std::vector<PageRange> pageRanges;
    wchar_t str2[1000];
    wcscpy_s(str2, 1000, str->Data());
    wchar_t* context;
    wchar_t* strRange = wcstok_s(str2, L" ,", &context);

    while (strRange != nullptr)
    {
        int dashIndex = wcscspn(strRange, L"-");

        // Just one page number
        if (dashIndex == wcslen(strRange))
        {
            int page = 0;
            if (Atoi(strRange, &page) &&
                page > 0 && page <= maxPageNumber)
            {
                PageRange pageRange = { page, page };
                pageRanges.push_back(pageRange);
            }
            else
            {
                return;
            }
        }
        // Two page numbers separated by a dash
        else
        {
            strRange[dashIndex] = '\0';
            int from = 0;
            int to = 0;

            if (Atoi(strRange, &from) &&
                Atoi(strRange + dashIndex + 1, &to) &&
                from > 0 && from <= maxPageNumber &&
                to > 0 && to <= maxPageNumber &&
                from <= to)
            {
                PageRange pageRange = { from, to };
                pageRanges.push_back(pageRange);
            }
            else
            {
                return;
            }
        }
        strRange = wcstok_s(nullptr, L" ,", &context);
    }

    if (pageRanges.size() == 0)
        return;

    // If we made it to this, the input string is valid
    this->pageMapping = ref new Vector<int>();

    // Define a mapping to page numbers
    for (unsigned int index = 0; index < pageRanges.size(); index++)
        for (int page = pageRanges[index].From; page <= pageRanges[index].To; page ++)
            pageMapping->Append(page);
}

bool CustomPageRange::Atoi(wchar_t * str, int * value)
{
    wchar_t overflow[10];
    return 1 == swscanf_s(str, L"%i %s", value, overflow, 10);
}

IVector<int>^ CustomPageRange::PageMapping::get()
{
    return pageMapping;
}

bool CustomPageRange::IsValid::get()
{
    return this->PageMapping != nullptr;
}
